### PR TITLE
[Clang] Fix definition of layout-compatible to ignore empty classes

### PR DESCRIPTION
**Pull request by MitalAshok**
Also changes the behaviour of `__builtin_is_layout_compatible`

None of the historic nor the current definition of layout-compatible classes mention anything about base classes (other than implicitly through being standard-layout) and are defined in terms of members, not direct members.

### CODE REVIEW

**Comment by dwblaikie** [comment]
How's this compare with other implementations clang is trying to be compatible with (gcc (in the normal clang driver mode) and msvc (in clang-cl mode))? Would this be an ABI break against them, or is this intended as an ABI fix to align better with them? Or some third option?

File: clang/lib/AST/DeclCXX.cpp [hunk 1]
```diff
@@ -561,6 +561,42 @@ void CXXRecordDecl::addedClassSubobject(CXXRecordDecl *Subobj) {
     data().StructuralIfLiteral = false;
 }
 
+const CXXRecordDecl *CXXRecordDecl::getStandardLayoutBaseWithFields() const {
+#ifndef NDEBUG
```
Comment (Endilll) [inline clang/lib/AST/DeclCXX.cpp:565]: We have `EXPENSIVE_CHECKS` macro to guard expensive checks.

File: clang/lib/AST/DeclCXX.cpp [hunk 2]
```diff
@@ -561,6 +561,42 @@ void CXXRecordDecl::addedClassSubobject(CXXRecordDecl *Subobj) {
...
+  {
+    assert(
+        isStandardLayout() &&
+        "getStandardLayoutBaseWithFields called on a non-standard-layout type");
```
Comment (Endilll) [inline clang/lib/AST/DeclCXX.cpp:569]: I think this assert should be on the default code path, not just for expensive checks.

File: clang/lib/AST/DeclCXX.cpp [hunk 3]
```diff
@@ -561,6 +561,42 @@ void CXXRecordDecl::addedClassSubobject(CXXRecordDecl *Subobj) {
...
+    unsigned NumberOfBasesWithFields = 0;
+    if (!field_empty())
+      ++NumberOfBasesWithFields;
+    std::set<const CXXRecordDecl *> UniqueBases;
```
Comment (Endilll) [inline clang/lib/AST/DeclCXX.cpp:573]: Have you considered `llvm::DenseSet`?

**Review by Endilll — COMMENTED** [review]
Thank you for working on this!

**Comment by MitalAshok** [comment]
@dwblaikie This patch will bring Clang in line with GCC and MSVC: https://godbolt.org/z/nj715zbsW

File: clang/lib/AST/DeclCXX.cpp (hunk 3 above)
Comment (MitalAshok) [reply clang/lib/AST/DeclCXX.cpp:573]: @Endilll `llvm::SmallPtrSet` seemed more appropriate here

**Comment by dwblaikie** [comment]
> @dwblaikie This patch will bring Clang in line with GCC and MSVC: https://godbolt.org/z/nj715zbsW

would the change be ABI incompatible with previous versions of clang? If so, then it'll need to be versioned in Clang's ClangABI handling, so that platforms (like Sony's Playstation, and Apple's platforms) that are only interested in Clang ABI compatibility can retain the old behavior.

@rjmccall might have an interest in whether this is an ABI thing Apple cares about
@pogo59 not sure who over at Sony's the best person to check on sony ABI things

**Comment by MitalAshok** [comment]
Clang only uses `Sema::isLayoutCompatible` in two places: to implement `__is_layout_compatible` and [Clang type safety checking attributes](https://clang.llvm.org/docs/AttributeReference.html#type-safety-checking), which warn if a pointer does not point to a layout compatible type. This only affects compiler warnings.

The type safety attributes were introduced by e4a5a90e8d6bc6a5ed0bf5f152a658c680b94867 and they were exposed as `__is_layout_compatible` by d5922cf72cc18a7ac9f7afd1941ee2f7773d8469 this February, so I don't think it was available in Clang 18

**Comment by MitalAshok** [comment]
https://github.com/llvm/llvm-project/blob/b0ae923ada836fa2c9114ac2c5afb39466f49fe0/clang/docs/ReleaseNotes.rst#L210-L212

Should be enough for ReleaseNotes, since not much has changed since Clang 18

**Comment by MitalAshok** [comment]
Ping

**Comment by MitalAshok** [comment]
I am also looking to cherry-pick this for Clang 19 so there won't be a difference between Clang 19 (when `__is_layout_compatible` was added) and Clang 20

File: clang/docs/ReleaseNotes.rst [hunk 4]
```diff
@@ -160,6 +160,9 @@ Bug Fixes in This Version
 Bug Fixes to Compiler Builtins
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+- ``__is_layout_compatible`` no longer requires the empty bases to be the same in two
+  standard-layout classes. It now only compares non-static data members.
```
Comment (rjmccall) [inline clang/docs/ReleaseNotes.rst:164]: Hmm.  I was under the impression that this feature had not been previously released; is that incorrect?

**Review by rjmccall — COMMENTED** [review]
The code changes look right to me.

File: clang/docs/ReleaseNotes.rst (hunk 4 above)
Comment (Endilll) [reply clang/docs/ReleaseNotes.rst:164]: That is correct, but the feature made into the 19 branch, whereas this fix did not (I hope we can backport it), hence a release note.

Comment (rjmccall) [reply clang/docs/ReleaseNotes.rst:164]: Is the plan to remove the release note from LLVM 20 if we're able to backport the fix?

Comment (Endilll) [reply clang/docs/ReleaseNotes.rst:164]: I'm not sure, but this indeed is a possible course of action.
If you have preferences how this should be handled, let us know.
Personally I'm more focused on getting this backported.

Comment (AaronBallman) [reply clang/docs/ReleaseNotes.rst:164]: We should backport these changes, drop the Clang 20 release note, and no need for a Clang 19.x release note to be added.

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Comment by Endilll** [comment]
@MitalAshok Let us know when you apply the changes suggested by Aaron, and consider this PR ready to be merged. Unfortunately, clock is ticking on backporting.

**Comment by MitalAshok** [comment]
@Endilll @AaronBallman Removed the release note and the tests have passed again. Would appreciate if someone else could start the backport process afterwards since I don't know how to do it.

**Pull request merged**